content/blog/eng/how-i-became-a-gitlab-contributor-to-fix-an-issue-in-our-projects-cicd-pipeline.md

```diff
@@ -0,0 +1,387 @@
+
+>ℹ️
+>
+>The complete history of my changes is tracked in the [merge request #3158](https://gitlab.com/gitlab-org/gitlab-runner/-/merge_requests/3158) on **GitLab Runner**.
+
+My contributions were merged to **GitLab** with version **14.5** and to **GitLab Runner** with version **14.8**, making me a contributor to their platform!
+
+{{< figure src="/images/how-i-became-a-gitlab-contributor-to-fix-an-issue-in-our-projects-cicd-pipeline/contributor.png" alt="My contributor badge on GitLab" class="image featured" >}}
+
+## Our final **CI/CD** pipeline
+
+At the end of this process, we were finally able to move away from **Docker Compose** while still having a **CI/CD** pipeline which didn't need workarounds or hacks to be fully functioning. With **GitLab** 14.8, our example pipeline can now be modified as it follows:
+
+```yaml
+job-1:
+    services:
+        - name: $MOCK_IMAGE:$COMMIT_ID
+          alias: company-1-mock
+          variables:
+            COMPANY_NAME: company-1
+            OTHER_VAR: ...
+        - name: $MOCK_IMAGE:$COMMIT_ID
+          alias: company-2-mock
+          variables:
+            COMPANY_NAME: company-1
+            OTHER_VAR: ...
+    script:
+        - npm run tests # Generic wrapper to test all the companies
+```
+
+If we consider the issues mentioned in our [first workaround](#a-first-workaround), we can see that with this solution we were able to fix them all:
+
+* ~~we can't define the `variables` as plain `YAML` properties as we did with **Docker Compose**~~ `variables` are now defined as plain `YAML` properties✅
+* ~~we still need to define a shared `variable` for each `service` at `job` level, meaning that `company-2-mock` can do `echo $COMPANY_1_ENVIRONMENT` and retrieve all the values for the other **mock server**, thus breaking the isolation between `services`~~ `variables` are now defined on a `service` basis, thus they're not shared anymore✅
+* ~~we need to manually add the original **Docker** `command` at the end of our overriden one, meaning that we need to keep it updated in case the base **Docker** image changes~~ we don't need to override the `entrypoint` and `command` for the `service` **Docker** image✅
+
+# Conclusions
+
+The goal of this post has been to show how an issue in everyday activities, coupled with a winning **opensource** culture, can be turned into an opportunity to grow up as a professional and to become a contributor to one of the world's biggest **DevOps** platforms. Usually we're somehow "limited" by our tooling when it comes to issues, and this means that we're kind of forced to find "hacky" workarounds to overcome them. By relying on widely supported **opensource** software, however, not only we can easily find a better solution to our problems, but we also have the chance to become part of the movement that enables our everyday tasks as IT professionals.
```

Comment:
The goal of this post was to show how a problem in daily tasks, coupled with a winning **opensource** culture, can be turned into an opportunity to grow as a professional and become a contributor to one of the largest **DevOps** platforms in the world. We're usually somehow "limited" by our tools when it comes to problems, and that means we're somehow forced to find "hacky" workarounds to overcome them. However, by relying on widely supported **opensource** software, not only can we easily find a better solution to our problems, but we also have the opportunity to become part of the movement that enables our daily tasks as IT professionals.